src/providers/openstack/configdrive.rs

```diff
@@ -0,0 +1,209 @@
+#[derive(Debug, Deserialize)]
+pub struct MetadataEc2JSON {
+    /// Local hostname.
+    pub hostname: Option<String>,
+    /// Instance ID.
+    #[serde(rename = "instance-id")]
+    pub instance_id: Option<String>,
+    /// Instance type.
+    #[serde(rename = "instance-type")]
+    pub instance_type: Option<String>,
+    /// Local IPV4.
+    #[serde(rename = "local-ipv4")]
+    pub local_ipv4: Option<String>,
+    /// Public IPV4.
+    #[serde(rename = "public-ipv4")]
+    pub public_ipv4: Option<String>,
+}
+
+/// Partial object for openstack `meta_data.json`
+#[derive(Debug, Deserialize)]
+pub struct MetadataOpenstackJSON {
+    /// SSH public keys.
+    pub public_keys: Option<HashMap<String, String>>,
+}
+
+/// OpenStack config-drive.
+#[derive(Debug)]
+pub struct OpenstackConfigDrive {
+    /// Path to the top directory of the mounted config-drive.
+    drive_path: PathBuf,
+    /// Temporary directory for own mountpoint (if any).
+    temp_dir: Option<TempDir>,
+}
+
+impl OpenstackConfigDrive {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        const TARGET_FS: &str = "iso9660";
+        let target = tempfile::Builder::new()
+            .prefix("afterburn-")
+            .tempdir()
+            .chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            TARGET_FS,
+            3,
+        )?;
+
+        let cd = OpenstackConfigDrive {
+            drive_path: target.path().to_owned(),
+            temp_dir: Some(target),
+        };
+        Ok(cd)
+    }
+
+    /// Return the path to the metadata directory.
+    fn metadata_dir(&self, platform: &str) -> PathBuf {
+        self.drive_path.clone().join(platform).join("latest")
+    }
+
+    /// Parse metadata attributes
+    ///
+    /// Metadata file contains a JSON object, corresponding to `MetadataEc2JSON`.
+    fn parse_metadata_ec2<T: Read>(input: BufReader<T>) -> Result<MetadataEc2JSON> {
+        serde_json::from_reader(input).chain_err(|| "failed parse JSON metadata")
+    }
+
+    /// Parse metadata attributes
+    ///
+    /// Metadata file contains a JSON object, corresponding to `MetadataOpenstackJSON`.
+    fn parse_metadata_openstack<T: Read>(input: BufReader<T>) -> Result<MetadataOpenstackJSON> {
+        serde_json::from_reader(input).chain_err(|| "failed parse JSON metadata")
+    }
+
+    /// The metadata is stored as key:value pair in ec2/latest/meta-data.json file
+    fn read_metadata(&self) -> Result<MetadataEc2JSON> {
+        let filename = self.metadata_dir("ec2").join("meta-data.json");
+        let file =
+            File::open(&filename).chain_err(|| format!("failed to open file '{:?}'", filename))?;
+        let bufrd = BufReader::new(file);
+        Self::parse_metadata_ec2(bufrd)
```

Comment:
According to the reference doc linked at the top:
> Do not rely on the presence of the EC2 metadata in the metadata API or config drive, because this content might be removed in a future release. For example, do not rely on files in the ec2 directory.

Can we move to get as many details as we can from `parse_metadata_openstack` instead (e.g. hostname and AZ), and also make `parse_metadata_ec2` failures not stopping the whole process?

Reply:
Yep, that makes sense. Was trying to align with the EC2 compatible metadata service we used in `openstack/network.rs`. I've shifted `hostname` to use from openstack config and use EC2 data as a backup.